uaclient/apt.py

```diff
@@ -22,36 +23,37 @@
 CA_CERTIFICATES_FILE = '/usr/sbin/update-ca-certificates'
 
 
-class InvalidAPTCredentialsError(RuntimeError):
-    """Raised when invalid token is provided for APT access"""
-    pass
-
-
-def valid_apt_credentials(repo_url, username, password):
+def assert_valid_apt_credentials(repo_url, username, password):
     """Validate apt credentials for a PPA.
 
     @param repo_url: private-ppa url path
     @param username: PPA login username.
     @param password: PPA login password or resource token.
 
-    @return: True if valid or unable to validate
+    @raises: UserFacingError for invalid credentials or timeout
+    @raises: ProcessExecutionError for unexpected errors
     """
     protocol, repo_path = repo_url.split('://')
     if not os.path.exists('/usr/lib/apt/apt-helper'):
-        return True   # Do not validate
+        return None
     try:
         util.subp(['/usr/lib/apt/apt-helper', 'download-file',
                    '%s://%s:%s@%s/ubuntu/pool/' % (
                        protocol, username, password, repo_path),
-                   '/tmp/uaclient-apt-test'],
-                  capture=False)  # Hide credentials from logs
-        return True
-    except util.ProcessExecutionError:
-        pass
+                   '/tmp/uaclient-apt-test'])
+    except util.ProcessExecutionError as e:
+        if e.exit_code == 100:
+            stderr = str(e.stderr)
+            if re.search(r'401\s+Unauthorized|HttpError401', stderr):
+                raise exceptions.UserFacingError(
+                    'Invalid APT credentials provided for %s' % repo_url)
+            elif re.search(r'Connection timed out', stderr):
+                raise exceptions.UserFacingError(
```

Comment:
Raising a UserFacingError for this means that we won't retry, but I think a connection timeout is something we should be retrying (because it could be a transient networking issue or temporary server outage).

Reply:
Oh, hold on, I think I have things mixed up here; we've already done all the retrying we would do in the util.subp call so disregard this.

Reply:
Hrm, in this specific case the timeout on apt-helper is 2 mins, this makes a  retry here crazy long.

Reply:
Ok, so we likely aren't going to retry this specific timeout case at this time. a user facing error will allow someone to ua enable esm again if they need to for the time being, I've filed #626 to capture the timeout length discussion to see if we think we should reduce on that timeout a bit.